### PR TITLE
Allow overriding of CSR accumulators with an `IndexLike`

### PR DESCRIPTION
**Pull request by thetorpedodog**
To support a custom indexer in TileDB-SOMA, this introduces a basic abstraction that is used to build the Indexer used by the CSR Accumulator.  This can then be easily overridden by an implementation's custom subclass to swap that out without having to duplicate substantial parts of the code.

The naming is, admittedly, not ideal; this is in part due to the naming of the things we're trying to abstract over (the Pandas `Index` type which has the `get_indexer` method).

---

This is the somacore counterpart of https://github.com/single-cell-data/TileDB-SOMA/pull/1728. It *is* entirely compatible with current code, though; once we’re positive that it fufills the needs of the tiledb-soma reindexer, we can merge it and release it and tiledbsoma installations will continue to work fine.

### CODE REVIEW

**Review by bkmartinjr — APPROVED** [review]
I did a code read, and this looks great to me.  Thanks for cleaning up some of the crufty code!

**Comment by beroy** [comment]
Plz fix the formattings

File: python-spec/src/somacore/query/types.py
Comment (thetorpedodog) [inline python-spec/src/somacore/query/types.py:29]: The reason we have this take the NDArray as the only parameter is so that we can use the function `pandas.Index` as an index factory. With this added, `pandas.Index` no longer works, which is why the format check is now failing.

**Comment by beroy** [comment]
@thetorpedodog can you plz merge this as my C++ reindexer is dependent on this?

**Review by beroy — APPROVED** [review]

**Comment by thetorpedodog** [comment]
I would like to wait until we get approval on the TileDB-SOMA side of the house. I am _pretty_ confident that this is going to be it, but I don’t want to cut a release prematurely. It will be very quick—after merging this, we can immediately make the tag and everything will be ready in a matter of minutes. (There will be no rush to release TileDB-SOMA; this is fully compatible with existing TileDB-SOMA.)

**Comment by thetorpedodog** [comment]
We’re good to go!

**Pull request merged**
